File: day03/zig/src/main.zig
const std = @import("std");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    var stdout_buf: [256]u8 = undefined;
    var stdout_writer = std.fs.File.stdout().writer(&stdout_buf);
    const stdout = &stdout_writer.interface;

    const file = try std.fs.cwd().openFile("../input.txt", .{});
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    var banks: std.ArrayList([]const u8) = .empty;
    defer banks.deinit(allocator);

    var line_iter = std.mem.splitScalar(u8, content, '\n');
    while (line_iter.next()) |line| {
        const trimmed = std.mem.trim(u8, line, &std.ascii.whitespace);
        if (trimmed.len > 0) {
            try banks.append(allocator, trimmed);
        }
    }

    const part1 = totalMaxJoltage(banks.items, 2);
    const part2 = totalMaxJoltage(banks.items, 12);

    try stdout.print("Part 1: {d}\n", .{part1});
    try stdout.print("Part 2: {d}\n", .{part2});
    try stdout.flush();
}

fn totalMaxJoltage(banks: []const []const u8, digits_to_select: usize) i64 {
    var total: i64 = 0;
    for (banks) |bank| {
        total += maxJoltageForBank(bank, digits_to_select);
    }
    return total;
}

fn maxJoltageForBank(bank: []const u8, digits_to_select: usize) i64 {
    var selected: [12]i64 = undefined;
    var start_index: usize = 0;

    for (0..digits_to_select) |position| {
        // Must leave (digits_to_select - position - 1) digits after this pick
        const end_index = bank.len - (digits_to_select - position - 1);
        const result = findMaxInRange(bank, start_index, end_index);
        selected[position] = result.max_digit;
        start_index = result.max_index + 1;
    }

    var joltage: i64 = 0;
    for (0..digits_to_select) |i| {
        joltage = joltage * 10 + selected[i];
    }
    return joltage;
}

const MaxResult = struct {
    max_digit: i64,
    max_index: usize,
};

fn findMaxInRange(bank: []const u8, start: usize, end: usize) MaxResult {
    var max_digit: i64 = -1;
    var max_index: usize = start;

    for (start..end) |i| {
        const digit: i64 = @intCast(bank[i] - '0');
        if (digit > max_digit) {
            max_digit = digit;
            max_index = i;
        }
    }

    return .{ .max_digit = max_digit, .max_index = max_index };
}

File: day03/rust/src/main.rs
use std::fs;

fn main() {
    let input = fs::read_to_string("../input.txt").expect("Failed to read input file");
    let banks: Vec<&str> = input.lines().filter(|l| !l.is_empty()).collect();

    println!("Part 1: {}", total_max_joltage(&banks, 2));
    println!("Part 2: {}", total_max_joltage(&banks, 12));
}

fn total_max_joltage(banks: &[&str], digits_to_select: usize) -> i64 {
    banks.iter().map(|bank| max_joltage_for_bank(bank, digits_to_select)).sum()
}

fn max_joltage_for_bank(bank: &str, digits_to_select: usize) -> i64 {
    let digits: Vec<i64> = bank.chars().map(|c| c.to_digit(10).unwrap() as i64).collect();
    let selected = select_max_digits(&digits, digits_to_select);
    selected.iter().fold(0i64, |acc, &digit| acc * 10 + digit)
}

fn select_max_digits(digits: &[i64], count: usize) -> Vec<i64> {
    // Greedy: for each position, pick the largest digit while leaving enough for remaining
    let mut result = Vec::with_capacity(count);
    let mut start_index = 0;

    for position in 0..count {
        // Must leave (count - position - 1) digits after this pick
        let end_index = digits.len() - (count - position - 1);
        let (max_digit, max_index) = find_max_in_range(digits, start_index, end_index);
        result.push(max_digit);
        start_index = max_index + 1;
    }

    result
}

fn find_max_in_range(digits: &[i64], start: usize, end: usize) -> (i64, usize) {
    let mut max_digit = -1;
    let mut max_index = start;

    for i in start..end {
        if digits[i] > max_digit {
            max_digit = digits[i];
            max_index = i;
        }
    }

    (max_digit, max_index)
}
